base/subarray.jl

```diff
@@ -461,3 +461,38 @@ function _indices_sub(i1::AbstractArray, I...)
 end
 
 has_offset_axes(S::SubArray) = has_offset_axes(S.indices...)
+
+# faster implementation of folding operations for "reversed" FastSubArray with stride1 == -1
+
+function reverse_fastsubarray_1(a)
+# here we assume a.stride1 == -1
+    fi, li = a.offset1 .- (firstindex(a), lastindex(a))
+    view(parent(a), li:fi)
+end
+
+function mapfoldl_impl(f::F, op::OP, init::I, a::FastSubArray) where {F,OP,I}
+    if a.stride1 == -1
+        b = reverse_fastsubarray_1(a)
+        @invoke mapfoldr_impl(f::F, FlipArgs(op)::FlipArgs{OP}, init::I, b::AbstractArray)
+    else
+        @invoke mapfoldl_impl(f::F, op::OP, init::I, a::AbstractArray)
+    end
+end
+
+function mapfoldr_impl(f::F, op::OP, init::I, a::FastSubArray) where {F,OP,I}
+    if a.stride1 == -1
+        b = reverse_fastsubarray_1(a)
+        @invoke mapfoldl_impl(f::F, FlipArgs(op)::FlipArgs{OP}, init::I, b::AbstractArray)
+    else
+        @invoke mapfoldr_impl(f::F, op::OP, init::I, a::AbstractArray)
+    end
+end
```

Comment:
```suggestion
function mapfoldl_impl(f::F, op::OP, nt, a::FastSubArray) where {F,OP}
    op′, itr′ = _xfadjoint(BottomRF(op), Generator(f, a))
    if itr′ isa FastSubArray && itr′.stride1 == -1
        return foldl_impl(op′, nt, _reverse_iter(reverse_fastsubarray_1(itr′)))
    else
        return foldl_impl(op′, nt, itr′)
    end
end

function mapfoldr_impl(f::F, op::OP, nt, a::FastSubArray) where {F,OP}
    op′, itr′ = _xfadjoint(BottomRF(FlipArgs(op)), Generator(f, a))
    if itr′ isa FastSubArray && itr′.stride1 == -1
        return foldl_impl(op′, nt, reverse_fastsubarray_1(itr′))
    else
        return foldl_impl(op′, nt, _reverse_iter(itr′))
    end
end
```
Perhaps this is better.

Reply:
Your proposal copies the logic from `mapfoldl_impl` and `mapfoldr_impl`. I didn't want to interfere with how these functions are actually implement for `AbstractArray` because any future change of that would have to be mirrored for `FastSubArray`. Do this convince you? Otherwise I change as you have suggested.

Reply:
>Does this convince you?

Make sense. I just think a `FlipArgs(FlipArgs(...))` is ugly and the current implement is wrong as `FlipArgs(op) !<: FlipArgs{OP}` if `op` isa constructor, e,g, `Pair`. 

In fact, when I switch to the above code, I increasingly feel that this is not a good idea.
If we just want to avoid boundcheck, then I think an overall fix for #48773 would be much better.
(If guys still dislike `@propagte_inbounds`, then we can add a `_foldl_impl` for `AbstractArray`.)

If we want more auto-vectorlization, I don't think `stride1 == -1` is that special. And we still need tools for other simdable `stride1`.

Reply:
> `FlipArgs(op) !<: FlipArgs{OP}` if `op` is a constructor, e,g, `Pair`

We absolutely have to avoid wrong code. However, I get
```
julia> using Base: FlipArgs
julia> op = Pair; FlipArgs(op) isa FlipArgs{typeof(op)}
true
julia> op = FlipArgs(+); FlipArgs(op) isa FlipArgs{typeof(op)}
true
```

> an overall fix for #48773 would be much better

I fully agree. I'd like to see what comes out of my [Discourse post](https://discourse.julialang.org/t/is-reduce-safe/95525). So far there has been no clear statement whether `@inbounds` is acceptable inside folding operations.

> If we want more auto-vectorization, I don't think `stride1 == -1` is that special. And we still need tools for other simdable `stride1`.

Is there auto-vectorization for non-contiguous data? That would be great. However, I would expect that even in that case the step value has to known at compile time. If that's true, then there is a limit to what we can do. I'm thinking that the -1 case is still fairly common. But again, a general solution would be much better.

Reply:
> We absolutely have to avoid wrong code. However, I get

My fault, didn't notice that `FlipArgs(op)` is not stable for constructor. Perhaps we should fix it.

>  I would expect that even in that case the step value has to known at compile time.

You are right, without compile time info LLVM would generate a simd branch for dense input and check if it's valid at runtime.
Since this only benifits simdable cases, it makes sense to me to to ask users to provide more static information if this does matter.
We already support range with static step in `Static.jl`, but that's not enough as we use the cached `stride1` in `getindex`.